flybywiresim/a32nx
# feat(ui): Notifications and Popups

Fixes #[issue_no]

## Summary of Changes

1. Notify the user when a setting change is made that requires an airplane reload. 
- Creates a Pop-Up that takes focus of the user's screen until dismissed (Only on MCDU)
- Creates a notification which persists for 10 seconds.

2. Notify the user to calibrate their throttles in the EFB
- Creates a notification that persists for 1 minute when the following conditions have been met:
  - Calibration has not been completed in the EFB
  - At least 10 minutes have passed
  - Physical throttle 1 axis has not been moved in that time
  - Flight is in a cruise phase (as determined by FWC)
  - LVR CLB message is showing on the PFD.
  - Throttle 1 is at a position above a point slightly below the lower bounds of the climb detent.

## Screenshots (if necessary)



https://user-images.githubusercontent.com/15316958/124350034-ceb78880-dc24-11eb-9773-28725a9ad143.mp4

Note: Notification styles have changed since WU5. The above video is outdated.

https://cdn.discordapp.com/attachments/747571985831297145/874942900390223872/nh0B9U0bCq.webm

## References

N/A, sim UX related PR.




## Additional context

Also added utility clases to A32NX_Utils.js (for JS) and @instruments/common (notification/popup.ts) for posterity.

## Code Examples

**React**
```
import { PopUp } from '@instruments/common/popup';
import { Notification } from '@instruments/common/notification';

...
const popup = new PopUp();
const notification = new Notification();
const yesFunc = () => {
    notification.showNotification({ message: 'Your notification here!' });
};
popup.showPopUp({ title: "YOUR TITLE HERE", message: "Your message here", style: "small"}, yesFunc, noFunc);
...
```

**HTML+JS**
```
const popup = new NXPopUp();
const notif = new NXNotif();
const yesFunc = () => {
    notif.showNotification({ message:"Your message here"});
};
popup.showPopUp({ title: "YOUR TITLE HERE", message: "Your message here", style: "small"}, yesFunc, noFunc);
```



Discord username (if different from GitHub): 2Cas#1022

## Testing instructions


**Please note if you are using the MS Store version or Steam version**

For settings reload notification:

1. Load the plane in any state
2. Change a MCDU option that requires a plane reload (KG/LB conversion or Keyboard Input)
3. Observe Pop-Up with Y/N options
4. When Pop-Up is cleared, a notification (w associated sound event) will display for 10 seconds on the HUD in the top center of the screen.

For calibration notification:

5. Delete any previous throttle calibration. Open File Explorer, go to %appdata%/Microsoft Flight Simulator/Packages/flybywire-aircraft-a320-neo/work and delete ThrottleConfiguration.ini. (N.B. For convenience, you don't actually need to delete this file, just move it out of the folder and back it up somewhere safe). 
6. Go to cruise (CRZ page on lower ECAM), making sure that AP is on and A/THRST is active. Set throttle either on the CLB detent, or a bit behind it, such that LVR CLB flashes on the PFD. Leave it in this state for at least 10 minutes, do not move the throttles.
7. A notification should pop up, and this will persist for a minute before disappearing. This notification will reappear every 5 minutes if the problem still exists.
8.  Enter the EFB and calibrate as per normal. On the next flight, the notification will not trigger even if the above steps are repeated, as long as a custom calibration has been set.

Known Issues:

* English only, no localisation possible for now.
* Default sim's Notifications can clash if enabled in the assistants options, and this behavior can be jarring at times.


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
